src/tests/pmemobj/res_pub/reserve_publish_test.cc

```diff
@@ -0,0 +1,526 @@
+    obj.act = &act[th * (messages_per_thread)];
+    obj.act_delete = &act_free[th * deletes_per_thread];
+    threads.push_back(
+        std::thread(&PMemObjReservePublishTest::ReservePublishDeferFreeInThread,
+                    this, std::ref(obj)));
+  }
+
+  /* Step 5 */
+  for (auto& th : threads) {
+    th.join();
+  }
+  threads.clear();
+
+  int remaining_space =
+      makeMaximumAllocations(pop, data_size, &act[total_messages]);
+
+  /* Step 6 */
+  int ret = pmemobj_publish(pop, act, total_messages + remaining_space);
+  ASSERT_EQ(ret, 0);
+
+  /* Step 7 */
+  ret = pmemobj_publish(pop, act_free, total_deletes);
+  ASSERT_EQ(ret, 0);
+
+  /* Step 8 */
+  pmemobj_close(pop);
+  pop = pmemobj_open(pool_path_.c_str(), LAYOUT_NAME);
+
+  /* Step 9 */
+  int free_count = makeMaximumAllocations(pop, data_size, &act[0]);
+  ASSERT_EQ(free_count, total_deletes);
+
+  /* Step 10 */
+  pmemobj_close(pop);
+
+  /* Step 11 */
+  ApiC::RemoveFile(pool_path_);
+  delete[] act;
+  delete[] act_free;
+}
+
+/**
+ * RESERVE_PUBLISH_DEFER_FREE_02
+ * Parameterized Test Case: Verifies that objects are not freed when the
+ * defer_free actions are cancelled
+ *  - the number of threads (n)
+ *  - the size of the objects to be reserved
+ *  - the number of objects to be reserved by each thread
+ * \test
+ *          \li \c Step1. Create the PMEMobj pool file
+ *          / SUCCESS function returns pointer to pool
+ *          \li \c Step2. Make reservations with pmemobj_reserve using n threads
+ *          / SUCCESS function returns handle to newly created reserved object
+ *          \li \c Step3. Populate the reserved object in the worker thread
+ *          \li \c Step4. Mark every 2nd object to be freed after publishing
+ *          with pmemobj_defer_free
+ *          \li \c Step5. Synchronize the threads
+ *          \li \c Step6. Publish the reserved objects with pmemobj_publish
+ *          / SUCCESS function returns 0
+ *          \li \c Step7. Cancel the free actions with pmemobj_cancel
+ *          \li \c Step8. Close and reopen the PMEMobj pool
+ *          \li \c Step9. Verify that none of the OID's were freed
+ *          \li \c Step10. Close the PMEMobj pool with pmemobj_close
+ *          \li \c Step11. Remove the PMEMobj pool
+ */
+TEST_P(PMemObjReservePublishParamTest, RESERVE_PUBLISH_DEFER_FREE_02) {
+  size_t max_messages = pool_size / data_size;
+  pobj_action* act = new pobj_action[max_messages];
+  size_t total_messages = nof_threads * messages_per_thread;
+  size_t deletes_per_thread = (messages_per_thread + 1) / 2;
+  size_t total_deletes = nof_threads * deletes_per_thread;
+  pobj_action* act_free = new pobj_action[total_deletes];
+
+  std::vector<std::thread> threads;
+
+  /* Step 1 */
+  pop = pmemobj_create(pool_path_.c_str(), LAYOUT_NAME, pool_size,
+                       S_IWRITE | S_IREAD);
+  ASSERT_TRUE(pop != nullptr) << pmemobj_errormsg();
+
+  for (int th = 0; th < nof_threads; th++) {
+    test_obj obj;
```

Comment:
This needs to be heap allocated, you are giving a new thread a stack-allocated value that lives only for the scope of one loop iteration.